### ISSUE TITLE
[BUG] ColumnTransformer passthrough doesn't always work 

### ISSUE THREAD

**Issue by davidgilbertson** [sktime/sktime#4785]
**Describe the bug**
When I have a `ColumnTransformer` with two transforms operating on the same column (one is `Id()` one is `WindowSummarizer`) then `ColumnTransformer` forgets to pass through the remainder of the columns.

**To Reproduce**
This code works as expected:
```python
import numpy as np
import pandas as pd
from sktime.transformations.compose import Id
from sktime.transformations.panel.compose import ColumnTransformer
from sktime.transformations.series.summarize import WindowSummarizer

df = pd.DataFrame(
    dict(
        A=np.random.normal(size=5),
        B=np.random.normal(size=5),
        C=np.random.normal(size=5),
    )
)

transformer= ColumnTransformer(
    [
        ("col1", Id(), ["A"]),
        (
            "col2",
            WindowSummarizer(
                lag_feature={"lag": [1, 2]},
                truncate="bfill",
                n_jobs=1,
            ),
            ["B"],
        ),
    ],
    remainder="passthrough",
)

print(transformer.fit_transform(df))

```

And outputs this (note column C is there):
```txt
        A  B_lag_1  B_lag_2       C
0 -0.1583  -1.5775  -1.5775 -0.6509
1 -0.4888  -1.5775  -1.5775 -0.9251
2  1.6321  -0.7883  -1.5775  0.1432
3 -0.4012  -1.1564  -0.7883  0.9755
4  1.0094   0.4105  -1.1564 -0.5998
```

But what I really want is to keep A, _and_ get some lags. If I change the `WindowSummarizer` to operate on column `A`:
```py
import numpy as np
import pandas as pd
from sktime.transformations.compose import Id
from sktime.transformations.panel.compose import ColumnTransformer
from sktime.transformations.series.summarize import WindowSummarizer

df = pd.DataFrame(
    dict(
        A=np.random.normal(size=5),
        B=np.random.normal(size=5),
        C=np.random.normal(size=5),
    )
)

transformer= ColumnTransformer(
    [
        ("col1", Id(), ["A"]),
        (
            "col2",
            WindowSummarizer(
                lag_feature={"lag": [1, 2]},
                truncate="bfill",
                n_jobs=1,
            ),
            ["A"],
        ),
    ],
    remainder="passthrough",
)

print(transformer.fit_transform(df))

```

Now the output doesn't have C any more:
```txt
   col1__A  col2__A_lag_1  col2__A_lag_2
0   0.6071         0.6071         0.6071
1  -0.0186         0.6071         0.6071
2  -0.6216        -0.0186         0.6071
3   0.3466        -0.6216        -0.0186
4   1.3375         0.3466        -0.6216
```

For context, I'm trying to see if I can get sktime to leave the column name of A as is, otherwise I would just do `"lag": [0, 1, 2]`. For even more context, this is because I want to pass column names to `HistGradientBoostingRegressor(categorical_features=...)` so at a high level I want to say "sktime, please pass the following columns through as is, and transform some other columns like so, and any column I don't mention pass that through".

I tested with sklearn's ColumnTransformer and couldn't get it to misbehave in the same way.

**Versions**
<details>

System:
    python: 3.10.8 (main, Oct 12 2022, 19:14:26) [GCC 9.4.0]
executable: /home/davidg/.virtualenvs/learning/bin/python
   machine: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
Python dependencies:
          pip: 23.1.2
       sktime: 0.19.1
      sklearn: 1.2.2
        numpy: 1.24.3
        scipy: 1.10.1
       pandas: 2.0.1
   matplotlib: 3.7.0
       joblib: 1.2.0
  statsmodels: 0.13.5
        numba: None
     pmdarima: 2.0.3
      tsfresh: None
   tensorflow: None
tensorflow_probability: None

</details>

**Comment by fkiraly** [comment]
@davidgilbertson, thanks for reporting!

I wasn't aware this transformer is still there - I would recommend you try out `ColumnEnsembleTransformer` from `transformations.compose`. This does not have a `remainder` argument.

The `ColumnTransformer` inherits directly from `sklearn` and may make assumptions such as a column may appear only once. The issue may well be present in the `sklearn` version.
We should probably look into implementing the functionality in `ColumnEnsembleTransformer` and deprecating `ColumnTransformer`.

Another way to do what you want that should work is sth like this:

```python
from sktime.transformations.compose import Id
from sktime.transformations.series.lag import Lag

t = Id()[["A", "C"]] * Lag([1, 2])["A"]
```

**Comment by fkiraly** [comment]
I was just looking at `ColumnEnsembleTransformer`, looks like a buggy mess - it slipped under the tests as it was not exported, it seems.

Either way, while we fix this, using `FeatureUnion` as above (e.g., with the dunders) should work

**Comment by davidgilbertson** [comment]
This seems to be a bug in sktime, not scikit-learn. With scikit-learn you can definitely transform a single column in multiple ways.

With this code:
```py
transformer = ColumnTransformer(
    [
        ("col1", Id(), ["A"]),
        ("col2", Id(), ["A", "B"]),
    ],
    remainder="passthrough",
)

print(transformer.fit_transform(df))

sk_transformer = SkLearnColumnTransformer(
    [
        ("col1", Normalizer(), ["A"]),
        ("col2", Normalizer(), ["A", "B"]),
    ],
    remainder="passthrough",
).set_output(transform="pandas")

print(sk_transformer.fit_transform(df))
```
I get this output (first block from sktime, second block from scikit-learn):
```txt
    col1__A   col2__A   col2__B
0  0.931024  0.931024  1.978871
1  1.762134  1.762134  2.835829
2  0.440897  0.440897  0.286572
3 -0.689826 -0.689826  1.686142
4  1.345185  1.345185  0.608733
   col1__A   col2__A   col2__B  remainder__C
0      1.0  0.425719  0.904856     -0.274296
1      1.0  0.527787  0.849376     -2.641460
2      1.0  0.838453  0.544973      2.617352
3     -1.0 -0.378652  0.925539     -0.252699
4      1.0  0.911058  0.412278     -0.827123
```

Regarding the dunders example, some user feedback from a newbie: I find them quite confusing and opaque. I much prefer code where I can search for the components to learn/remember what it's doing. With `[]` and `*` and `**` it's a bit difficult to unravel what's going on. I was excited when I found the cheatsheet but I've read this five times and still don't know what it's saying:
![image](https://github.com/sktime/sktime/assets/4443482/d1415adb-74f0-4198-a691-0de62ebc5a96)

"chaining to secondary input of another estimator" - what's 'secondary input'? And even though I've spent a week reading the docs, I'm still struggling to work out what's going to modify X and what's going to modify y, with the `*` and `**` it's even more confusing, and you can't search for those in the docs. Also, the results of the dunder method aren't typed, so I lose autocomplete. And my final complaint is that the resulting column names are huge, like `TransformerPipeline_2__lag_2__A` which makes troubleshooting difficult.

OK back to your suggestion :) Yes I think that's good. (You used `*`, but I assume you meant `+`). Given `passthrough` doesn't work, that leaves `ColumnTransformer` out of action, so just manually doing a `FeatureUnion` with `Id()` for all the other columns gets the job done. It's just a pity there's no `verbose_feature_names_out=False` option as there is in scikit-learn, because now I've lost all my original column names (my workaround is a `FunctionTransformer` that fixes all the column names).

**Comment by fkiraly** [comment]
> You used `*`, but I assume you meant `+`

Yes, you are right. Corrected the example above.

> And my final complaint is that the resulting column names are huge, like `TransformerPipeline_2__lag_2__A` which makes troubleshooting difficult.

This could be solved in the following way:

* dunder combinations of `+` and `[ ]` like the above are parsed to automatically become a `ColumnEnsembleTransformer` (i.e., additions of transformers that do or do not have `[ ] ` applied)
* the `ColumnEnsembleTransformer` is bugfixed to do what you want

In combination that would fix the column name issue...

**Comment by fkiraly** [comment]
> Regarding the dunders example, some user feedback from a newbie: I find them quite confusing and opaque. I much prefer code where I can search for the components to learn/remember what it's doing. With `[]` and `*` and `**` it's a bit difficult to unravel what's going on. I was excited when I found the cheatsheet but I've read this five times and still don't know what it's saying

Ok, maybe you can help us improving this bit. Let me try to explain:

* in transformations, the "most important" input is the variable `X`. Some transformers also take `y` into account
* in forecasters, the "most important" input is the variable `y`. Some forecasters also take `X` into account
* in classifiers, the "most important" input is the variable `X`.

So, doing `my_trafo * est` should apply `my_trafo` to `X` if `est` is a classifier or transformer, and to `y` if `est` is a forecaster.

Whereas `my_trafo ** est` should apply `my_trafo` to `y` if `est` is a transformer, and to `X` if `est` is a forecaster.

**Comment by fkiraly** [comment]
> Also, the results of the dunder method aren't typed, so I lose autocomplete.

Can you kindly explain what you mean by this? What is the "should happen" state?

**Comment by fkiraly** [comment]
> It's just a pity there's no `verbose_feature_names_out=False` option as there is in scikit-learn,

Interesting - in this re-work, I included a similar variable to flatten or not the `MultiIndex`: https://github.com/sktime/sktime/pull/4789

What do you think about the following solution?

Have a variable `out_var_names`, which can be:

* `"flat"`, which produces names such as `"Estimator1__A"`, `"Estimator2__B"`
* `"multiindex"`, which produces a `MultiIndex` tuple such as `("Estimator", "A")`, `("Estimator2", "B")`
* `"original"`, which produces names such as `"A"`, `"B"` - if this leads to non-unique names, it defaults to `"flat"` for any non-unique name instead

**Comment by davidgilbertson** [comment]
Heh, I _thought_ I got `*` and `**` but your explanation was news to me. I think I get it now, although I question the first half of this: "Whereas `my_trafo ** est` should apply `my_trafo` to `y` if `est` is a transformer, and to `X` if `est` is a forecaster." - isn't it invalid if both the left and right side of `**` are transformers?

My understanding so far:
![image](https://github.com/sktime/sktime/assets/4443482/41a423b1-b769-4eb0-906a-3d6088a35a6e)

As for typing, yes let me elaborate. To be clear, I'm speaking of static typing (what PyCharm/VS Code use for autocomplete), as opposed to dynamic analysis (what Jupyter does). The return value of, say, `__rpow__` doesn't have a type defined, so IDE's don't know what it is:
![image](https://github.com/sktime/sktime/assets/4443482/a7e78c18-fb60-47f5-841f-943de01cc378)

But with the longhand approach, they know and I get autocomplete, etc.
![image](https://github.com/sktime/sktime/assets/4443482/4dd0a0b2-c765-4cbf-836f-2b74628fa912)

And useful warnings for unfilled parameters:
![image](https://github.com/sktime/sktime/assets/4443482/93eea3b6-1dd3-4156-a420-052532b1bde2)

Which makes development much faster especially since I've not yet memorized every method and parameter :)

About `verbose_feature_names_out`. Ideally it would be the same as scikit-learn to minimize friction, but to comment on "original", I think that's fine given the extra `multiindex` setting. I would normally have one `ColumnTransformer` that does all of the 'in-place' transformations of columns, where I want the name to stay the same. Followed by more complex transforms that might add more columns that require new names.

**Comment by fkiraly** [comment]
> but your explanation was news to me

Hm, how can we improve this? There is some explanation at the end of the tutorial `03_transformers`, but I see that it does not explain the `**`.

The notebook this is in for forecasters is `03b_forecasting_etc` - perahps that should precede the cheat sheet? Or is that still not too clear?

> isn't it invalid if both the left and right side of `**` are transformers?

Currently, yes, although one could imagine that it should apply the transformer to the `y` of the right hand transformer.

**Comment by fkiraly** [comment]
> My understanding so far:
![image](https://github.com/sktime/sktime/assets/4443482/41a423b1-b769-4eb0-906a-3d6088a35a6e)

That's right!

This is also a nice table - should we add it to the cheat sheet, plus some explanation? Would much appreciate a PR, this looks like a good visual summary!

**Comment by fkiraly** [comment]
> About `verbose_feature_names_out`. Ideally it would be the same as scikit-learn to minimize friction, but to comment on "original", I think that's fine given the extra `multiindex` setting. 

Yes, indeed - the `sklearn` design only allows two settings because it is boolean. Side note: that's precisely the reason why I prefer to use string parameters (and not bool) even if it has only two settings, who knows whether later it will make sense to add a third.

> I would normally have one `ColumnTransformer` that does all of the 'in-place' transformations of columns, where I want the name to stay the same. Followed by more complex transforms that might add more columns that require new names.

So, you'd rather have the current `ColumnEnsembleTransformer` (repaired) *replace* `ColumnTransformer` with the same name?

**Comment by davidgilbertson** [comment]
By "name to stay the same" I meant the column names, I have no strong opinion about what the class names are.

I'm keen to contribute something, but right now I'm still trying to work out if I can actually get my current project (doing manual recursion with a Gradient Boosting regressor) working in sktime, and it looks like it's just not possible. I think the discussion over at https://github.com/sktime/sktime/issues/3224 covers the problems I'm facing, although I haven't read the whole thing.

So either I'm missing something, or I need to try and write my own custom component, or I can't use sktime. I guess I could just iterate over my instances and have sktime in an inner loop, or maybe I can get `pooling="global"` to get decent results by tweaking the regressor properties (I have a few thousand instances).

What's your opinion, should I pause my efforts to try and use sktime and come back in six months or a year? Is there something I'm missing?

**Comment by fkiraly** [comment]
> What's your opinion, should I pause my efforts to try and use sktime and come back in six months or a year? Is there something I'm missing?

I assume you mean this with your project?
https://github.com/sktime/sktime/issues/4776

Re 4776, I had hoped that @danbartl, the original author behind current `make_reduction` logic, would show up on 4776 and help out.

But how about this "deal": I prioritize looking at 4776 for half a day or so, and you contribute sth? E.g., filling out the user registration form at sktime.net, or doing the above?

**Comment by davidgilbertson** [comment]
Sure, I can fill out a form :)
